### ISSUE TITLE
[MPMD-246] Output details of processing errors

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-pmd-plugin#583]
**[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel)** opened **[MPMD-246](https://issues.apache.org/jira/browse/MPMD-246?redirect=false)** and commented

To help debugging processing errors (or any error in PMD), maven should be able to output the processing errors with exception stacktraces.

Currently, only the number of the problematic files are displayed.

Furthermore, PMD usually adds the processing errors to the XML report. But the XML report that is generated with the maven pmd plugin doesn't contain the errors.

---

**Issue Links:**
- [MPMD-244](https://issues.apache.org/jira/browse/MPMD-244) Maven PMD plugin fails but no reason is given for the failure

2 votes, 4 watchers

**Comment by jira-importer** [comment]
**[Juan Martín Sotuyo Dodero](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jmsotuyo)** commented

Bare in mind PMD's XML report, starting from PMD 6.0.0, has an explicit XML Schema, and will not only provide stacktraces for Processing Errors, but Config Errors are rendered too, and will gain relevance in upcoming releases.

This may be a good chance to include these too, so that the Maven Plugin produces reports in line with PMD; or simply let PMD create the XML report rather than build it from the Maven plugin itself.

**Comment by jira-importer** [comment]
**[Ryan Bard](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rbard)** commented

I am having this problem as well.  I've investigated what was going wrong in my files (what java code reproduces the problem, I don't know what is going wrong in the plugin code).

It is something to do with processing java code that has an anonymous inner class overriding a method that infers it's generic type from the parameter.

I have a writeup of this info and a small example here: https://github.com/RyanBard/pmd-plugin-bug

EDIT: It's not inferring the type from the method param, but it is important to have an inferred type in the method parameters (no parameters didn't reproduce the problem for me).

**Comment by jira-importer** [comment]
**[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel)** commented

[Ryan Bard](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rbard) Thanks for the testcase. This is actually a problem in PMD itself - I've verified, it is already fixed with the upcoming version 6.0.0 (to be released tomorrow).

This issue here is about the problem, that it's virtually impossible at the moment, to investigate the processing error (you currently don't get a stacktrace out of PMD when running via m-pmd-p).

The reason, why you don't see the processing error when executing PMD on the command line is this: It's the `auxclasspath` option, which is enabled by default by the maven-pmd-plugin. It improves the detection rate of PMD, since it can the also take type information into account and thus avoiding false positives. Maven is a great help here, since it assembles the runtime classpath for you (the auxclasspath contains all dependencies of the project). It can however trigger some bug in PMD.
You can run maven in debug mode, to see, which auxclasspath is used:

```
$ ./mvnw clean verify -X
...
[INFO] --- maven-pmd-plugin:3.8:pmd (pmd) @ example1 ---
...
[DEBUG] Using aux classpath: [/home/andreas/PMD/source/pmd-it/MPMD-246-pmd-plugin-bug/example1/target/classes, /home/andreas/PMD/source/pmd-it/MPMD-246-pmd-plugin-bug/common/target/common-1.0-SNAPSHOT.jar]
...
```

With that, you can run it on the command line:

```
~/PMD/binaries/pmd-bin-5.8.1/bin/run.sh pmd -d example1/src -f text -R pmdrules.xml -auxclasspath example1/target/classes:common/target/common-1.0-SNAPSHOT.jar
```

And this now also fails with a processing error. Adding `-V` to the command line gives you then the stacktrace, which is

```
Caused by: java.lang.IllegalArgumentException: No generic parameter by name T on class 
	at net.sourceforge.pmd.lang.java.typeresolution.typedefinition.JavaTypeDefinition.getGenericType(JavaTypeDefinition.java:103)
...
```

**Comment by jira-importer** [comment]
**[Ryan Bard](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rbard)** commented

[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel), thank you for the information!

I've worked around the problem by setting skipPmdError to true and just creating a named class instead of anonymous-inner-classes in those cases in my project's code.

It was going to be really tempting to look at the 6.0.0 fix and maybe try to back-port it to 5.8.2 this weekend, however, if 6.0.0 is about to release, I'll just wait and use 6.0.0.

Thanks again for giving me some tools to better find out what is going on if I run into issues again in the future.

**Comment by jira-importer** [comment]
**[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel)** commented

With 3.9.0 a new parameter has been added for PMD, which is by default enabled:

Name: **renderProcessingErrors** (property = "pmd.renderProcessingErrors")

Also render processing errors into the HTML report.
Processing errors are problems, that PMD encountered while executing the rules.
It can be parsing errors or exceptions during rule execution.
Processing errors indicate a bug in PMD and the information provided help in
reporting and fixing bugs in PMD.
